edi_oca/models/edi_exchange_record.py

```diff
@@ -0,0 +1,572 @@
+
+    def notify_action_complete(self, action, message=None):
+        """Notify current record that an edi action has been completed.
+
+        Implementers should take care of calling this method
+        if they work on records w/o calling edi_backend methods (eg: action_send).
+
+        Implementers can hook to this method to do something after any action ends.
+        """
+        if message:
+            self._notify_related_record(message)
+
+        # Trigger generic action complete event on exchange record
+        event_name = f"{action}_complete"
+        self._trigger_edi_event(event_name)
+        if self.related_record_exists:
+            # Trigger specific event on related record
+            self._trigger_edi_event(event_name, target=self.record)
+
+    def _notify_related_record(self, message, level="info"):
+        """Post notification on the original record."""
+        if not self.related_record_exists or not hasattr(
+            self.record, "message_post_with_view"
+        ):
+            return
+        self.record.message_post_with_view(
+            "edi_oca.message_edi_exchange_link",
+            values={
+                "backend": self.backend_id,
+                "exchange_record": self,
+                "message": message,
+                "level": level,
+            },
+            subtype_id=self.env.ref("mail.mt_note").id,
+        )
+
+    def _trigger_edi_event_make_name(self, name, suffix=None):
+        return "on_edi_exchange_{name}{suffix}".format(
+            name=name,
+            suffix=("_" + suffix) if suffix else "",
+        )
+
+    def _trigger_edi_event(self, name, suffix=None, target=None, **kw):
+        """Trigger a component event linked to this backend and edi exchange."""
+        name = self._trigger_edi_event_make_name(name, suffix=suffix)
+        target = target or self
+        target._event(name).notify(self, **kw)
+
+    def _notify_done(self):
+        self._notify_related_record(self._exchange_status_message("process_ok"))
+        self._trigger_edi_event("done")
+
+    def _notify_error(self, message_key):
+        self._notify_related_record(
+            self._exchange_status_message(message_key),
+            level="error",
+        )
+        self._trigger_edi_event("error")
+
+    def _notify_ack_received(self):
+        self._notify_related_record(self._exchange_status_message("ack_received"))
+        self._trigger_edi_event("done", suffix="ack_received")
+
+    def _notify_ack_missing(self):
+        self._notify_related_record(
+            self._exchange_status_message("ack_missing"),
+            level="warning",
+        )
+        self._trigger_edi_event("done", suffix="ack_missing")
+
+    def _notify_ack_received_error(self):
+        self._notify_related_record(
+            self._exchange_status_message("ack_received_error"),
+        )
+        self._trigger_edi_event("done", suffix="ack_received_error")
+
+    @api.model
+    def _search(
```

Comment:
Someone have a look at the changes here, please.

Reply:
I have a suggestion here, please try it:

```

    @api.model
    def _search(self, domain, offset=0, limit=None, order=None, access_rights_uid=None):
        query = super()._search(
            domain,
            offset=offset,
            limit=limit,
            order=order,
            access_rights_uid=access_rights_uid,
        )

        if self.env.is_system():
            # restrictions do not apply to group "Settings"
            return query

        # TODO highlight orphaned EDI records in UI:
        #  - self.model + self.res_id are set
        #  - self.record returns empty recordset
        # Remark: self.record is @property, not field

        if query.is_empty():
            return query
        orig_ids = list(query)
        ids = set(orig_ids)
        result = []
        model_data = defaultdict(
            lambda: defaultdict(set)
        )  # {res_model: {res_id: set(ids)}}
        for sub_ids in self._cr.split_for_in_conditions(ids):
            self._cr.execute(
                """
                            SELECT id, res_id, model
                            FROM "%s"
                            WHERE id = ANY (%%(ids)s)"""
                % self._table,
                dict(ids=list(sub_ids)),
            )
            for eid, res_id, model in self._cr.fetchall():
                if not model:
                    result.append(eid)
                    continue
                model_data[model][res_id].add(eid)

        for model, targets in model_data.items():
            if not self.env[model].check_access_rights("read", False):
                continue
            recs = self.env[model].browse(list(targets))
            missing = recs - recs.exists()
            if missing:
                for res_id in missing.ids:
                    _logger.warning(
                        "Deleted record %s,%s is referenced by edi.exchange.record %s",
                        model,
                        res_id,
                        list(targets[res_id]),
                    )
                recs = recs - missing
            allowed = (
                self.env[model]
                .with_context(active_test=False)
                ._search([("id", "in", recs.ids)])
            )
            for target_id in allowed:
                result += list(targets[target_id])
        if len(orig_ids) == limit and len(result) < len(orig_ids):
            extend_query = self._search(
                domain,
                offset=offset + len(orig_ids),
                limit=limit,
                order=order,
                access_rights_uid=access_rights_uid,
            )
            extend_ids = list(extend_query)
            result.extend(extend_ids[: limit - len(result)])
        # Restore original ordering
        result = [x for x in orig_ids if x in result]
        if set(orig_ids) != set(result):
            # Create a virgin query
            query = self.browse(result)._as_query()
        return query
```

Reply:
Thanks for the suggestion!